### ISSUE TITLE
Controlling accumulation dtype for `dot_general` and Pallas `dot`

### ISSUE THREAD

**Issue by jon-chuang** [google/jax#17612]
Triton `dot` out_dtype default is `float32` (see [here](https://github.com/openai/triton/blob/08c16589573621fcb8cd5a9c3b8a0537077f876d/python/triton/language/core.py#L988))

Currently, we do not lower any `out_dtype`, either based on precision or other API knobs in the wrapper API `lax.dot_general`.

Hence, performance will suffer.

Note that for whatever reason (precision or otherwise), flash_attn accumulates elements in float32 (see [here](https://github.com/Dao-AILab/flash-attention/blob/d0032700d1c7a1353a3a8f2fadfbc73b2dc6b5dc/csrc/flash_attn/src/kernel_traits.h#L26))

**Comment by jon-chuang** [comment]
Near duplicate: https://github.com/google/jax/issues/15683

**Comment by sharadmv** [comment]
I think the temporary solution here is to add an `out_type` to `pl.dot` and pass that into `lax.dot(preferred_element_type=...)`.

**Comment by jon-chuang** [comment]
Yes, I did that locally. Worked somewhat well except that I needed to force the outdtype conversion with `dot(..., out_dtype=my_dtype).astype(my_dtype)`.

Would be nice to remove the extra verbosity of `.astype(my_dtype)`
